### ISSUE TITLE
Performance differences between @jit and @vectorize on arrays

### ISSUE THREAD

**Issue by pitrou** [numba/numba#1223]
The following functions:

```
def py_rel_diff(a, b):
    return 2 * (a - b) / (a + b)

rel_diff = vectorize(nopython=True)(py_rel_diff)
rel_diff_jit = jit(nopython=True)(py_rel_diff)
```

give different performance results on simple input arrays (1d). This issue will try to find out why.

**Comment by pitrou** [comment]
First, we must disable the "safe" alloc in NRT (i.e. the memset() call that fills the allocated array with a magic debug value).

**Comment by pitrou** [comment]
The second observation is that removing the division can make the @jit version faster: this is because type inference infers a different result type. For example np.add(int32, int32) returns int32, while the "+" operator, as inferred by Numba, returns int64 on the same inputs. This makes a large difference when LLVM is able to vectorize the loop.

**Comment by pitrou** [comment]
The third observation is that Python "/" (as used in the @vectorize code) checks for zero division, which can prevent vectorization. Numpy's np.divide (or np.true_divide) doesn't.

**Comment by pitrou** [comment]
The fourth observation is that some small change in the loop generation for `@vectorize` can allow LLVM to vectorize better. I don't know yet if the change is in all cases.

**Comment by pitrou** [comment]
The last remaining slowdown for `float64` is due to division wrongly checking for 0 in array expressions. This will be solved by PR #1769.
